packages/oauth-helper/src/index.ts

```diff
@@ -0,0 +1,468 @@
+  verifyStateParam: (now: Date, state: string) => Promise<string | undefined>;
+}
+
+// State object structure
+interface StateObj {
+  now: Date;
+  metadata?: string;
+}
+
+// default implementation of StateStore
+class ClearStateStore implements StateStore {
+  private stateSecret: string;
+  // todo: remove need for default string value here
+  public constructor(stateSecret: string | undefined) {
+    if (stateSecret == null) {
+      throw new Error('You must provide a State Secret to use the built-in state store');
+    }
+    this.stateSecret = stateSecret;
+  }
+
+  // TODO: Question, why do the params below need to have types definied
+  // instead of getting those types from StateStore interface
+  public generateStateParam(now: Date, metadata?: string): Promise<string> {
+    console.log('clear store generate state param');
+    const state = sign({ metadata, now: now.toJSON() }, this.stateSecret);
+    return new Promise<string>((resolve) => {
+      resolve(state);
+    });
+  }
+  public verifyStateParam(now: Date, state: string): Promise<string | undefined> {
+    // decode the state using the secret
+    const decoded: StateObj = verify(state, this.stateSecret) as StateObj;
+
+    // return metadata string
+    return new Promise<string | undefined>((resolve) => {
+      resolve(decoded.metadata);
+    });
+  }
+}
+
+interface InstallationStore {
+  storeInstallation: (installation: Installation) => Promise<void>;
+  fetchInstallation: (query: InstallationQuery) => Promise<Installation>;
+}
+
+interface DevDatabase {
+  [key: string]: Installation;
+}
+const devDB: DevDatabase = {};
+
+// Default Install Store. Should only be used for development
+class MemoryInstallStore implements InstallationStore {
+
+  public storeInstallation(installation: Installation): Promise<void> {
+    console.log('Storing Access Token. Please use a real Installation Store for production!');
+    // db write
+    devDB[installation.team.id] = installation;
+    return new Promise<void>((resolve) => {
+      resolve();
+    });
+  }
+
+  public fetchInstallation(query: InstallationQuery): Promise<Installation> {
+    console.log('Retrieving Access Token from DB. Please use a real Installation Store for production!');
+    // db read
+    const item = devDB[query.teamId];
+    return new Promise<Installation>((resolve) => {
+      resolve(item);
+    });
+  }
+}
+
+// Needs to have all the data from OAuthV2Access result and OAuthAccess
+// result. This is a normalized shape.
+interface Installation {
+  team: {
+    id: string;
+    name: string;
+  };
+  enterprise?: {
+    id: string;
+    name: string;
+  };
+  bot?: {
+    token: string;
+    scopes: string[];
+    id?: string;
+    userId: string;
+  };
+  user: {
+    token?: string;
+    scopes?: string[];
+    id: string;
+  };
+  incomingWebhook?: {
+    url: string;
+    channel: string;
+    configurationUrl: string;
+  };
+  appId: string | undefined;
+  tokenType?: string;
```

Comment:
I added `tokenType` because I use it in `callbackSuccess` below to determine if I should do the redirect